File: gcj/gcj2011/round2/c/c.l.not-finished.pl
use strict;
use utf8;
use Data::Dumper;
use bignum;

&main();

sub getPrimes
{
	my ($N, $primes) = @_;

#	my %no;
	my @no;
	
	#ignore 1
	for my $i (2..$N)
	{
		if ($i == 2)
		{
			push (@$primes, $i);
		}
		elsif ($i % 2 != 0 && !defined $no[$i])
		{
			push(@$primes, $i);
			my $num = $i;
			while($num <= $N)
			{
				$no[$i] = 1;
				$num += $i;
			}
		}
	}
}

sub main
{
	my $max = 1000000;
	my $Nmax = $max * $max;
	my @primes;
	getPrimes($max, \@primes);

	my %clues;
	$clues{2} = 1;
	for my $prime (@primes)
	{
#		print "prime = $prime.\n";
		my $num = $prime * $prime;
		while($num < $Nmax)
		{
			$clues{$num} = 1;
			$num *= $prime;
		}
	}
	
	my @sorted = sort keys %clues;

	my %rank;
	for my $i (0..$#sorted)
	{
		$rank{$sorted[$i]} = $i;
	}

	my $T = <STDIN>;
	for my $i (0..$T - 1)
	{
		my ($N) = split(' ', <STDIN>);
		print "Case #" . ($i + 1) . ": ";

		
	}

}

File: gcj/gcj2011/round2/c/c.l.not-finished.cpp
#include <cstdio>
#include <fstream>
#include <stdio.h>
#include <vector>
#include <algorithm>
#include <functional>
#include <set>
#include <map>
using namespace std;

typedef long long ll;

void getPrimes(ll n, vector<ll>&primes)
{
	vector<ll>::iterator itr;
	map<ll, int> isNotPrime;

	for (ll i = 2; i < n; i++)
	{
		if (i == 2)
		{
			primes.push_back(i);
		}
		
		else if(i % 2 != 0 && !isNotPrime[i])
		{
			//printf("prime = %lld\n", i);
			primes.push_back(i);
			ll num = i;
			while(num <= n)
			{
				isNotPrime[num] = 1;
				num += i;
			}
		}
	}
}

int main(int ARGC, char *ARGV[])
{
	ll Nmax = 1000000000000;
	ll max = 1000001;
	//ll Nmax = 10000000000;
	//ll max = 100000;
	vector<ll>primes;
	getPrimes(max, primes);
	printf("end.\n");

	vector<ll>::iterator itr;
	set<ll>keys;
	vector<ll>keys2;
	for (itr = primes.begin(); itr != primes.end(); itr++)
	{
		ll prime = *itr;
		ll num = prime * prime;
		while (num <= Nmax)
		{
			//keys.insert(num);
			keys2.push_back(num);
			num *= prime;
		}
	}
	//keys.insert(2);
	keys2.push_back(2);

	std::sort(keys2.begin(), keys2.end());
	printf("end.\n");
	printf("size = %lld\n", keys2.size());

	std::ifstream ifs( ARGV[1] );

	int testcaseNum;
	ifs >> testcaseNum;
	for (int t = 0; t < testcaseNum; t++)
	{
		ll N;
		ifs >> N;
		ll ans;

		while(1)
		{
			set<ll>::iterator pos = keys.find( N );
			if (pos == keys.end())
			{
				N--;
			}
			else
			{
				ans = *pos + 1;
			}
		}

		printf("Case #%d: %d\n", t + 1, ans);
		

	}

	return 0;
}
